nautilus: mds: add configurable snapshot limit

mchangir opened the pull request:
backport tracker: https://tracker.ceph.com/issues/42441

---

backport of https://github.com/ceph/ceph/pull/30710
parent tracker: https://tracker.ceph.com/issues/41209

this backport was staged using ceph-backport.sh version 15.0.0.6950
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

smithfarm commented:
Also please note that there is a newer version of `ceph-backport.sh` in master which addresses the "GitHub API deprecation" warning issue.

mchangir commented:
> This does not cherry-pick cleanly for me:
> 
> ```
> ceph-backport.sh: Attempting to cherry pick 1 commits from https://github.com/ceph/ceph/pull/30710 into local branch wip-42441-nautilus
> ceph-backport.sh: Running "git cherry-pick -x" on c15905cd977 mds: add configurable snapshot limit
> + git cherry-pick -x c15905cd9774a28d07d9a1151e18feb89d024d97
> error: could not apply c15905cd977... mds: add configurable snapshot limit
> hint: after resolving the conflicts, mark the corrected paths
> hint: with 'git add <paths>' or 'git rm <paths>'
> hint: and commit the result with 'git commit'
> + set +x
> ceph-backport.sh: ERROR: Cherry pick failed
> ceph-backport.sh: Next, manually fix conflicts and complete the current cherry-pick
> ceph-backport.sh: Then re-run the script
> ```
> 
> @mchangir Please append to your commit message the following:
> 
> ```
> Conflicts:
>         src/common/options.cc
> ```
> 
> And, if you see fit, include an explanation of the conflict and what you did to resolve it.
> 
> Also, please rebase to get rid of the merge commit.

* Added the Conflicts section and reason.
* Rebased PR to do away with merge commit.

smithfarm commented:
@mchangir Can you fetch the upstream repo to get the latest nautilus, and then rebase on top of that? GitHub is showing a conflict in `src/mds/Server.cc`.

mchangir commented:
I hope this works.

mchangir commented:
> @mchangir Can you fetch the upstream repo to get the latest nautilus, and then rebase on top of that? GitHub is showing a conflict in `src/mds/Server.cc`.

Please re-check.
I hope this works.

mchangir commented:
> @mchangir looks like the change to `MDSRank.cc` was lost: https://github.com/ceph/ceph/pull/30710/files#diff-ea4df8dd62d8f79137676b9f21d58633

merged the change into MDSRank.cc

batrick commented:
retest this please

batrick commented on src/common/options.cc:
```diff
@@ -8155,6 +8155,18 @@ std::vector<Option> get_mds_options() {
+    Option("mds_task_status_update_interval", Option::TYPE_FLOAT, Option::LEVEL_DEV)
+     .set_default(2.0)
+     .set_description("task status update interval to manager")
+     .set_long_description("interval (in seconds) for sending mds task status to ceph manager"),
```
This option is not part of the original PR. Looks like an accidental addition during conflict resolution.

mchangir replied:
removed this spurious diff

smithfarm commented:
@mchangir Also, the "Conflicts" section seems to have disappeared? See https://github.com/ceph/ceph/pull/33295#issuecomment-588823386

ajarr commented:
@mchangir please update PR

ajarr commented on qa/tasks/cephfs/test_snapshots.py:
```diff
@@ -443,3 +446,89 @@ def test_multimds_hardlink(self):
+
+    def create_dir_and_snaps(self, dir_name, snaps):
+        self.mount_a.run_shell(["mkdir", dir_name])
+        
```
remove white spaces

smithfarm replied:
@ajarr - this is a cherry-pick of  c15905c which has this whitespace, so I don't think we should be removing it here.

mchangir replied:
found them ... and removed

mchangir replied:
found them ... and removed

ajarr replied:
@smithfarm ok
@mchangir leave them as is.

ajarr commented on qa/tasks/cephfs/test_snapshots.py:
```diff
@@ -443,3 +446,89 @@ def test_multimds_hardlink(self):
+    def test_mds_max_snaps_per_dir_with_increased_limit(self):
+        """
+        Test the newly introudced option named mds_max_snaps_per_dir
+        First create 101 directories and ensure that the 101st directory 
```
remove whitespace

smithfarm replied:
@ajarr - this is a cherry-pick of c15905c which has this whitespace, so I don't think we should be removing it here.

mchangir replied:
found them ... and removed

smithfarm replied:
@mchangir In that case, can you open a PR removing them from master as well? Otherwise the whitespace difference can cause cherry-pick conflicts in the future.

mchangir replied:
> @mchangir In that case, can you open a PR removing them from master as well? Otherwise the whitespace difference can cause cherry-pick conflicts in the future.

@smithfarm FYI - #33826

mchangir commented:
uploading updated PR

smithfarm commented:
@mchangir Please add the conflicts section to the commit message as requested in https://github.com/ceph/ceph/pull/33295#pullrequestreview-358798125

mchangir commented:
> @mchangir Please add the conflicts section to the commit message as requested in [#33295 (review)](https://github.com/ceph/ceph/pull/33295#pullrequestreview-358798125)

@smithfarm 
The conflicts had arose due to the PR being based on an old nautilus HEAD.
The Conflicts section is no longer pertinent to the PR with the updated nautilus branch HEAD.

ajarr commented:
> > @mchangir Please add the conflicts section to the commit message as requested in [#33295 (review)](https://github.com/ceph/ceph/pull/33295#pullrequestreview-358798125)
> 
> @smithfarm
> The conflicts had arose due to the PR being based on an old nautilus HEAD.
> The Conflicts section is no longer pertinent to the PR with the updated nautilus branch HEAD.

I still see the conflict when cherry-picking c15905cd97  on latest nautilus
 ```
git status
On branch nautilus
Your branch is up to date with 'origin/nautilus'.

You are currently cherry-picking commit c15905cd97.
  (fix conflicts and run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

Changes to be committed:

	modified:   qa/tasks/cephfs/test_snapshots.py
	modified:   src/mds/MDSRank.cc
	modified:   src/mds/Server.cc
	modified:   src/mds/Server.h

Unmerged paths:
  (use "git add <file>..." to mark resolution)

	both modified:   src/common/options.cc
```

gregsfortytwo commented:
@mchangir I think you're misunderstanding; the "Conflicts" section records which files had conflicts in the cherry-pick, not whether there will be conflicts on the merge (which, if there are any, would be recorded in the merge commit).
So in this case, the cherry-pick has a conflict in options.cc since the options file was not the same between Nautilus and Octopus.

Modulo that change, this looks fine assuming it tests correctly.

gregsfortytwo commented:
Running tests at
http://pulpito.front.sepia.ceph.com:80/gregf-2020-03-13_20:56:54-fs-wip-greg-testing-nautilus-313-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com:80/gregf-2020-03-13_21:00:59-multimds:basic-wip-greg-testing-nautilus-313-distro-basic-smithi/

Once those run let's merge if they look good.

ajarr commented:
Same make check error as seen in https://github.com/ceph/ceph/pull/33877#issuecomment-599462189
